### PR TITLE
JSON-LD Design Patterns

### LINKED ISSUE #590: Use id/type instead of @id/@type
Schema.org, Social Web / ActivityStreams, Annotation WG will all adopt this.
Manu, danbri, tantek, etc all agree.

There will be a best practices document published from W3C that states this

### LINKED ISSUE #1131: If property can ever have multiple values, require a JSON array
From this thread:
> https://groups.google.com/forum/?hl=en#!topic/iiif-discuss/ihQ3nHanh8Y

The proposal is to force all properties that can have multiple values to always be an array, in all APIs.
In the Presentation API, this would be: `label, metadata, description, thumbnail, attribution, rights, logo, type, viewingHint, related, rendering, service, seeAlso, within`

This is done in JSON-LD by adding `@container: @set` into the context document for those properties.

Note however that JSON-LD 1.0 currently prevents us from doing it for the values of LanguageMaps:
> https://github.com/json-ld/json-ld.org/issues/407
> https://github.com/IIIF/iiif.io/issues/755

But this is fixed in 1.1 (c.f. #1192) 

An example from @workergnome:
> http://evil-manifests.davidnewbury.com/iiif/garden/everything_a_list.json

### LINKED ISSUE #1284: Pattern for when to use URI string or JSON object
Reviewing the situations when we use URIs as strings rather than as the `id` of a resource encoded as a JSON object, I propose the following design pattern:

If a URI could be treated as a value in an enumerated list, such that the context could assign a short form to it, then the URI _MUST_ be given as a string.  Conversely, if this is not the case and the number of resources is either unbounded or impossible to enumerate, it _MUST_ be given in the object form with at least an `id` and `type`. All enumerations defined by IIIF specifications _MUST_ be encoded in the respective context as easy-to-remember-and-use strings.

There are a countable number of renderingHints, renderingDirections, choiceHints, timeModes, __profiles__, features, and types, which must thus all be strings. There is an uncountable number of logos, licenses, thumbnails, canvases, images, etc. which thus must be given as objects.

### PR DESCRIPTION
**Pull request by azaroth42**
Closes #1284 ; Closes #1131  ; Closes #590

### CODE REVIEW

File: source/api/annex/notes/design_patterns.md [hunk 1]
```diff
@@ -112,11 +114,64 @@ IIIF specifications define the functionality that can be expected to work and ho
...
+These design patterns have been selected in order to maximize these features in the resulting APIs.
+
+### 3.1. Use Native Data Types
+
+Native data types in JSON are used to ensure that applications do not have to do type coercion. If the value should be an integer in the model, then the JSON representation is as a JSON number.  Similarly, if the value is an ordered list of resources, then a JSON array will be used, mapping to an rdf:List.
+
+### 3.2. Avoid Special Characters
+
+Special characters are avoided in the names of classes and properties. Any character other than those in the range 0-9, a-z, A-Z and the underscore character are removed in the mapping to JSON-LD, if present in the name of a class or property. This is to ensure that applications developed in object oriented programming languages can treat these values as class and property names internally.
+s
```
Comment (zimeon) [inline source/api/annex/notes/design_patterns.md]: stray "s"?

File: source/api/annex/notes/design_patterns.md [hunk 2]
```diff
@@ -112,11 +114,64 @@ IIIF specifications define the functionality that can be expected to work and ho
...
+
+  * Valid JSON and JSON-LD, following the restrictions of both specifications ;
+  * Understandable by developers without special training or documentation ;
+  * Usable in a variety of application environments to ensure the widest adoption.
+
+These design patterns have been selected in order to maximize these features in the resulting APIs.
+
+### 3.1. Use Native Data Types
+
+Native data types in JSON are used to ensure that applications do not have to do type coercion. If the value should be an integer in the model, then the JSON representation is as a JSON number.  Similarly, if the value is an ordered list of resources, then a JSON array will be used, mapping to an rdf:List.
```
Comment (zimeon) [inline source/api/annex/notes/design_patterns.md]: * _"a JSON number (unquoted)"_ <- add (unquoted) to make it really clear?
 * s/will be/is/ to be consistent
 * add back ticks around `rdf:List`?

File: source/api/annex/notes/design_patterns.md [hunk 3]
```diff
@@ -112,11 +114,64 @@ IIIF specifications define the functionality that can be expected to work and ho
...
+
+### 3.2. Avoid Special Characters
+
+Special characters are avoided in the names of classes and properties. Any character other than those in the range 0-9, a-z, A-Z and the underscore character are removed in the mapping to JSON-LD, if present in the name of a class or property. This is to ensure that applications developed in object oriented programming languages can treat these values as class and property names internally.
+s
+JSON-LD defines two special keys in the data: `@id` (the URI of the resource) and `@type` (the class of the resource).  These are aliased to `id` and `type` respectively to avoid the use of the `@` symbol.
+
+### 3.3. Use Consistent Naming Conventions
+
+Even within the limited range of characters available, there are additional design patterns to make it easier to remember.
```
Comment (zimeon) [inline source/api/annex/notes/design_patterns.md]: * Delete "Even" or perhaps reorder to "Additional design patterns are used make names easier to remember while keeping to the limited range of characters."?
 * s/it/names/ ?

**Comment by zimeon** [comment]
Modulo a few minor quibbles, great job on this @azaroth42

File: source/api/annex/notes/design_patterns.md [hunk 4]
```diff
@@ -112,11 +114,64 @@ IIIF specifications define the functionality that can be expected to work and ho
...
+
+#### 3.4.1. Always Use Arrays for Multiple Value Properties 
+
+If a property can ever have multiple values, it will always be an array even if it has only one value. This means that code can always iterate over the values without checking first to see if it's an array.  Thus we use `"language": [ "en" ]` and not `"language": "en"`.
+
+#### 3.4.2. Use JSON Objects for Referenced Resources
+
+If the set of referenced resources is not unbounded or otherwise impossible to enumerate, then the resource will be described with a JSON object that has at least the `id` and `type` properties. This ensures ease of processing as the resources will always be of the same type, rather than some being just the URI from `id` as a string and others being a JSON object. The inclusion of `type` makes it easier to build object models in code, as the class to instantiate is given by the description rather than needing to be inferred from the property.  Thus we use `"thumbnail": {"id": "https://example.org/images/logo.jpg", "Image"}` and not `"thumbnail": "https://example.org/images/logo.jpg"`.
```
Comment (mikeapp) [inline source/api/annex/notes/design_patterns.md]: "type": "Image"

Comment (zimeon) [reply source/api/annex/notes/design_patterns.md]: Well spotted!

**Comment by azaroth42** [comment]
Explicit :+1: for @zimeon's change to predominantly my branch (should probably be a process point)

**Comment by jpstroop** [comment]
Agree, this is great!

**Comment by azaroth42** [comment]
@tomcrane, you can merge :)

**Pull request merged**